README.md
![image](images/microchip.jpg) 

# MATLAB LVMC dsPIC33CK256MP508 LX34050 FOC

## INTRODUCTION
This document describes the setup requirements for running the LX34050 inductive position sensor Based FOC algorithm, using MATLAB/Simulink and dsPIC33CK Low Voltage Motor Control (LVMC) Board.

## 2.	SUGGESTED DEMONSTRATION REQUIREMENTS
### 2.1 MATLAB Model Required for the Demonstration
-  MATLAB model can be cloned or downloaded as zip file from the Github repository ([link](https://github.com/microchip-pic-avr-solutions/matlab-lvmc-dspic33ck256mp508-foc-inductive-position-sensor)).

### 2.2	Software Tools Used for Testing the MATLAB/Simulink Model
1.	MPLAB X IDE and IPE (v6.05)
2.	XC16 compiler (v2.00)
3.	MATLAB R2023a
4.	Required MATLAB add-on packages
    -	Simulink
    -	Simulink Coder
    -	MATLAB Coder
    -	Embedded Coder (v7.9)
    -	MPLAB Device blocks for Simulink (v3.50.35)
    - Motor Control Blockset (v1.5)

> **_NOTE:_**
>The software used for testing the model during release is listed above. It is recommended to use the version listed above or later versions for building the model.

### 2.3	Hardware Tools Required for the Demonstration
- dsPIC33CK Low Voltage Motor Control (LVMC) Development Board ([DM330031](https://www.microchip.com/en-us/development-tool/DM330031))
- 24V Power Supply ([AC002013](https://www.microchipdirect.com/dev-tools/AC002013)) 
- 24V, 3-Phase Brushless DC Permanent Magnet Hurst Motor ([AC300020](https://www.microchip.com/en-us/development-tool/AC300020))
- [LX34050](https://www.microchip.com/en-us/product/LX34050) Static Sensor Board
- Adapter Board 
- Target Conductor

> **_NOTE:_**
>All items listed under this section Hardware Tools Required for the Demonstration are available at [microchip DIRECT](https://www.microchipdirect.com/).

## 3. HARDWARE SETUP
<p style='text-align: justify;'>This section describes hardware setup required for the demonstration.</p>

1. Remove the back cover of the short hurst motor and attach the sensor wheel to the shaft as shown in the figure below.

    <img src="images/front_side.PNG" height = 50% width = 50%>

2. Screw the sensor board and connect the phase winding wires and sensor wires.  
<img src="images/wire_con.PNG" height = 45% width = 45%>   <img src="images/wire_conn.PNG" height = 33% width = 33%> 

3. The final set up is as shown below:
![image](images/actual_setup.png) 

4. <p style='text-align: justify;'> Plug in the 24V power supply to connector J1 provided on the dsPIC33CK LVMC Board. Alternatively, the Inverter Board can also be powered through Connector J2.</p>

    <p align="left" >
    <img  src="images/har3.png"></p>

5. <p style='text-align: justify;'> The board has an onboard programmer 'PICkit™ On Board (PKoBv4)', which can be used for programming or debugging the dsPIC33CK256MP508. To use an on-board programmer, connect a micro-USB cable between Host PC and Micro USB connector J13 provided on the dsPIC33CK LVMC Board.</p>

    <p align="left" >
    <img  src="images/har4.png"></p>

6. <p style='text-align: justify;'>	Alternatively, the device can also be programmed using the programmer/debugger (MPLAB® PICkit™ 4 In-Circuit Debugger - PG164140) by interfacing it through connector J10 of the dsPIC33CK LVMC Board, as shown below. Ensure that the programmer is oriented correctly before proceeding.</p>

    <p align="left" >
    <img  src="images/har5.png"></p>

## 4.	BASIC DEMONSTRATION
<p style='text-align: justify;'> Follow the below instructions step-by-step, to set up and run the motor control demo application:</p>

1. Launch MATLAB (refer the section [“2.2 Sofware Tools Used for Testing the MATLAB/Simulink Model"](#22-software-tools-used-for-testing-the-matlabsimulink-model)).</p> 
2. Open the folder downloaded from the repository, in which MATLAB files are saved (refer the section ["2.1 MATLAB Model Required for the Demonstration"](#21-matlab-model-required-for-the-demonstration)).

    <p align="left" >
    <img  src="images/dem2.png"></p>

3.	<p style='text-align: justify;'> Double click and open the .m file. This .m file contains the configuration parameter for the motor and board. By default, the .m file is configured to run Hurst 075 motor and LVMC board. Run the file by clicking the <b>“Run”</b> icon and wait till all variables gets loaded on the <b>‘Workspace’</b> tab.

    <p align="left">
      <img  src="images/dem3.png"></p>
    </p>

4.	<p style='text-align: justify;'>Double click on the FOC Simulink model.

    <p align="left">
      <img  src="images/simfilepic.png"></p>
    </p>

5.	<p style='text-align: justify;'>This opens the FOC Simulink model as shown below. Click on the <b>"Run"</b> icon to start the simulation.

    <p align="left">
      <img  src="images/dem5.png"></p>
    </p>
6.	<p style='text-align: justify;'>To plot the simulation result, <b>Data Inspector</b> is used (refer to figure below). To observe the additional signals, log them as required. Alternatively, normal Simulink Scope can be used to plot the signals.

    <p align="left">
      <img  src="images/dem6.png"></p>
    </p>
    
7.	<p style='text-align: justify;'>From this Simulink model an MPLAB X project can be generated, and it can be used to run the PMSM motor using LVMC board. <p style='text-align: justify;'>To generate the code from the Simulink model, go to the <b>"MICROCHIP"</b> tab, and enable the tabs shown in the figure below. 

    <p align="left">
      <img  src="images/dem7.png"></p>
    </p>

8.	<p style='text-align: justify;'>	To generate the code and run the motor, click on <b>‘Build Model’ or ‘Clean Build Model’</b> option under the <b>“Microchip”</b> tab. This will generate the MPLAB X project from the Simulink model and program the dsPIC33CK256MP508 device.

    <p align="left">
      <img  src="images/dem8.png"></p>
    </p>

9.	<p style='text-align: justify;'>After completing the process, the <b>‘Operation Succeeded’</b> message will be displayed on the <b>‘Diagnostics Viewer’</b>.

    <p align="left">
      <img  src="images/dem9.png"></p>
    </p>

10.	<p style='text-align: justify;'>If the device is successfully programmed, <b>LED- LD10 and LD11</b> will be blinking.

11.	<p style='text-align: justify;'> To Run the motor, press the push button <b>SW1</b>.

    <p align="left">
      <img  src="images/dem11.png"></p> 
    </p>

12.	The motor speed can be varied using the potentiometer (labeled <b>“POT1”</b>).

    <p align="left">
      <img  src="images/dem12.png"></p>
    </p>

13.	Press the push button <b>SW1</b> to stop the motor.

    <p align="left">
      <img  src="images/dem13.png"></p>
    </p>

## 5.	DATA VISUALIZATION USING MOTOR CONTROL BLOCKSET (MCB) HOST MODEL
<p style='text-align: justify;'>The FOC model comes with the initialization required for data visualization using Motor Control Blockset Host Model (MCB Host Model). The MCB Host Model is a Simulink model which facilitates data visualization through the UART Serial Interface. 

1.	<p style='text-align: justify;'>To establish serial communication with the host PC, connect a micro-USB cable between the host PC and the dsPIC33CK LVMC Board (J13 connector). This interface is used for programming as well. (Alternatively, a micro-USB cable can be connected from the host PC to the J6 connector of the dsPIC33CK LVMC Board to establish the serial communication).
    <p align="left">
      <img  src="images/host1.png"></p>
    </p>

2. Ensure that the FOC model is programmed and running as described under section ["4. Basic Demonstration"](#4-basic-demonstration) by following steps 1 through 13.

3. <p style='text-align: justify;'>Open the MCB Host model and double click on the <b>“Serial Setup”</b> block. Then select the appropriate COM port connected to the hardware from the drop-down menu and set the baud rate as 921659. Please note that the same baud rate is required to be chosen in the FOC model (the baud rate can be viewed on the <b>“UART Configuration”</b> block in the <b>“LVMC Board Template”</b>).

    <p align="left">
      <img  src="images/host3.png"></p>
    </p>

4.	<p style='text-align: justify;'>Open the <b>“UART_Rx”</b> subsystem to configure the COM port. This can be done by configuring the <b>“Host Serial Receive”</b> block of the “UART_Rx” subsystem. Ensure to select the same COM port configured in step 3. 

    <p align="left">
      <img  src="images/host4.png"></p>
    </p>

5.	<p style='text-align: justify;'>Click the run icon of the MCB Host model to open the scope window and monitor the signals.

    <p align="left">
      <img  src="images/host5.png"></p>
    </p>

6.	<p style='text-align: justify;'>In the figure below, one example is shown where two signals (measured and reference speeds) have been plotted.

    <p align="left">
      <img  src="images/host6.png"></p>
    </p>


## 	6. REFERENCES:
For more information, refer to the following documents or links.
1. Field Oriented Control (FOC) of a Three-Phase BLDC Motor Using a Rotary Inductive Position Sensor 
([AN4764](https://ww1.microchip.com/downloads/aemDocuments/documents/MCU16/ApplicationNotes/ApplicationNotes/AN4764-FOC-of-a-3-Phase-BLDC-Motor-Using-a-Rotary-Inductive-Position-Sensor-DS00004764.pdf))
2.	dsPIC33CK LVMC Board User’s Guide ([DS50003297](http://ww1.microchip.com/downloads/en/DeviceDoc/DS50002927a.pdf)) 
3.	[MPLAB® X IDE installation](https://microchipdeveloper.com/mplabx:installation)
4.	[MPLAB® XC16 Compiler installation](https://microchipdeveloper.com/mplabx:installation)
5.  [Motor Control Blockset](https://in.mathworks.com/help/mcb/)
6.  [MPLAB Device Blocks for Simulink :dsPIC, PIC32 and SAM mcu](https://in.mathworks.com/matlabcentral/fileexchange/71892-mplab-device-blocks-for-simulink-dspic-pic32-and-sam-mcu)

 










changelog.md
# matlab-lvmc-dspic33ck256mp508-foc-inductive-position-sensor v1.0.0
### Release Highlights
This is the first version of MATLAB Simulink model for field oriented control of PMSM using inductive position sensor on dsPIC33CK LVMC board.
The model is set up for running Hurst075 (AC300020) motor.


### Features Added\Updated





matlab-lvmc-dspic33ck256mp508-inductive-position-sensor/mcb_pmsm_foc_inductive_position_dsPIC33_data.m
%% ************************************************************************
% Model         :   PMSM Field Oriented Control
% Description   :   Set Parameters for PMSM Field Oriented Control
% File name     :   mcb_pmsm_foc_inductive_position_dsPIC33_data.m
% Copyright 2023 The MathWorks, Inc.

%% Simulation Parameters

%% Set PWM Switching frequency
PWM_frequency 	= 20e3;    %Hz          // converter s/w freq
T_pwm           = 1/PWM_frequency;  %s  // PWM switching time period

%% Set Sample Times
Ts          	= T_pwm;        %sec        // simulation time step for controller
Ts_simulink     = T_pwm/2;      %sec        // simulation time step for model simulation
Ts_motor        = T_pwm/2;      %Sec        // Simulation sample time
Ts_inverter     = T_pwm/2;      %sec        // simulation time step for average value inverter
Ts_speed        = 30*Ts;        %Sec        // Sample time for speed controller

%% Set data type for controller & code-gen
dataType = fixdt(1,16,14);    % Fixed point code-generation
dataType2 = fixdt(1,16,12);    % Fixed point code-generation

%% System Parameters
% Set motor parameters

%Short Hurst Motor (Uncomment while using this motor from line below)
pmsm.model  = 'Hurst075';          %           // Manufacturer Model Number
pmsm.sn     = 'DMB0224C10002';         %           // Manufacturer Model Number
pmsm.p = 5;                     %           // Pole Pairs for the motor
pmsm.Rs = 2.8322;                %Ohm        // Stator Resistor
pmsm.Ld = 0.00242345;               %H          // D-axis inductance value
pmsm.Lq = 0.0023278;               %H          // Q-axis inductance value
pmsm.Lav= (pmsm.Ld+pmsm.Lq)/2;     %H          // avg inductance value
pmsm.Ke = 7.5678;                  %Bemf Const	// Vline_peak/krpm
pmsm.Kt = 0.03;                %Nm/A       // Torque constant
pmsm.J = 5.5e-6;     %Kg-m2      // Inertia in SI units
pmsm.B = 15.2987e-6;     %Kg-m2/s    // Friction Co-efficient
pmsm.I_rated= 1.16*sqrt(2);     %A      	// Rated current (phase-peak)
pmsm.QEPSlits = 1000;           %           // QEP Encoder Slits
pmsm.N_max  = 2500;             %rpm        // Max speed
pmsm.FluxPM     = (pmsm.Ke)/(sqrt(3)*2*pi*1000*pmsm.p/60); %PM flux computed from Ke
pmsm.T_rated    = (3/2)*pmsm.p*pmsm.FluxPM*pmsm.I_rated;   %Get T_rated from I_rated

%% Inductive Position Sensor offset parameters
sine.offset = 7760 ;
cosine.offset = 7848 ;

%% Inverter parameters

inverter.model         = 'LVMC';         % 		// Manufacturer Model Number
inverter.sn            = 'INV_XXXX';         		% 		// Manufacturer Serial Number
inverter.V_dc          = 24;       					%V      // DC Link Voltage of the Inverter
inverter.ISenseMax     = 21.85; 					%Amps   // Max current that can be measured
inverter.I_trip        = 10;                  		%Amps   // Max current for trip
inverter.Rds_on        = 1e-3;                      %Ohms   // Rds ON
inverter.Rshunt        = 0.01;                     %Ohms    // Rshunt
inverter.R_board       = inverter.Rds_on + inverter.Rshunt/3;  %Ohms
inverter.MaxADCCnt     = 4095;      				%Counts // ADC Counts Max Value
inverter.invertingAmp  = -1;                        % 		//Non inverting current measurement amplifier
inverter.deadtime      = 1e-6;
inverter.OpampFb_Rf    = 4.02e3;                    %Ohms  //Opamp Feedback resistance for current measurement
inverter.opampInput_R  = 532;                       %Ohms  //Opamp Input resistance for current measurement
inverter.opamp_Gain    = inverter.OpampFb_Rf/inverter.opampInput_R; %Opamp Gain used for current measurement

%% Derive Characteristics
pmsm.N_base = 3125;%mcb_getBaseSpeed(pmsm,inverter); %rpm // Base speed of motor at given Vdc

%% PU System details // Set base values for pu conversion
PU_System = mcb_SetPUSystem(pmsm,inverter);

%% Controller design // Get ballpark values!
% Get PI Gains
PI_params = mcb.internal.SetControllerParameters(pmsm,inverter,PU_System,T_pwm,Ts,Ts_speed);

%Updating delays for simulation
PI_params.delay_Currents    = int32(Ts/Ts_simulink);
PI_params.delay_Position    = int32(Ts/Ts_simulink);
PI_params.delay_Speed       = int32(Ts_speed/Ts_simulink);
PI_params.delay_Speed1       = (PI_params.delay_IIR + 0.5*Ts)/Ts_speed;

matlab-lvmc-dspic33ck256mp508-inductive-position-sensor/dsPIC33CK256MP508_IPS_LVMC_sh.m
%% ************************************************************************
% Model         :   PMSM Field Oriented Control
% Description   :   Set Parameters for PMSM Field Oriented Control
% File name     :   mcb_pmsm_foc_inductive_position_dsPIC33_data.m
% Copyright 2023 The MathWorks, Inc.

%% Simulation Parameters

%% Set PWM Switching frequency
PWM_frequency 	= 20e3;    %Hz          // converter s/w freq
T_pwm           = 1/PWM_frequency;  %s  // PWM switching time period

%% Set Sample Times
Ts          	= T_pwm;        %sec        // simulation time step for controller
Ts_simulink     = T_pwm/2;      %sec        // simulation time step for model simulation
Ts_motor        = T_pwm/2;      %Sec        // Simulation sample time
Ts_inverter     = T_pwm/2;      %sec        // simulation time step for average value inverter
Ts_speed        = 30*Ts;        %Sec        // Sample time for speed controller

%% Set data type for controller & code-gen
dataType = fixdt(1,16,14);    % Fixed point code-generation
dataType2 = fixdt(1,16,12);    % Fixed point code-generation
Typecast_BEMF = fixdt(1,16,10);
Typecast_ci = fixdt(1,16,11);
Typecast_di = fixdt(1,16,10);

%% System Parameters
% Set motor parameters

%Short Hurst Motor (Uncomment while using this motor from line below)
pmsm.model  = 'Hurst075';          %           // Manufacturer Model Number
pmsm.sn     = 'DMB0224C10002';         %           // Manufacturer Model Number
pmsm.p = 5;                     %           // Pole Pairs for the motor
pmsm.Rs = 2.8322;                %Ohm        // Stator Resistor
pmsm.Ld = 0.00242345;               %H          // D-axis inductance value
pmsm.Lq = 0.0023278;               %H          // Q-axis inductance value
pmsm.Lav= (pmsm.Ld+pmsm.Lq)/2;     %H          // avg inductance value
pmsm.Ke = 7.5678;                  %Bemf Const	// Vline_peak/krpm
pmsm.Kt = 0.03;                %Nm/A       // Torque constant
pmsm.J = 5.5e-6;     %Kg-m2      // Inertia in SI units
pmsm.B = 15.2987e-6;     %Kg-m2/s    // Friction Co-efficient
pmsm.I_rated= 1.16*sqrt(2);     %A      	// Rated current (phase-peak)
pmsm.QEPSlits = 1000;           %           // QEP Encoder Slits
pmsm.N_max  = 2500;             %rpm        // Max speed
pmsm.FluxPM     = (pmsm.Ke)/(sqrt(3)*2*pi*1000*pmsm.p/60); %PM flux computed from Ke
pmsm.T_rated    = (3/2)*pmsm.p*pmsm.FluxPM*pmsm.I_rated;   %Get T_rated from I_rated

%% Inverter parameters

inverter.model         = 'LVMC';         % 		// Manufacturer Model Number
inverter.sn            = 'INV_XXXX';         		% 		// Manufacturer Serial Number
inverter.V_dc          = 24;       					%V      // DC Link Voltage of the Inverter
inverter.ISenseMax     = 21.85; 					%Amps   // Max current that can be measured
inverter.I_trip        = 10;                  		%Amps   // Max current for trip
inverter.Rds_on        = 1e-3;                      %Ohms   // Rds ON
inverter.Rshunt        = 0.01;                     %Ohms    // Rshunt
inverter.R_board       = inverter.Rds_on + inverter.Rshunt/3;  %Ohms
inverter.MaxADCCnt     = 4095;      				%Counts // ADC Counts Max Value
inverter.invertingAmp  = -1;                        % 		//Non inverting current measurement amplifier
inverter.deadtime      = 1e-6;
inverter.OpampFb_Rf    = 4.02e3;                    %Ohms  //Opamp Feedback resistance for current measurement
inverter.opampInput_R  = 532;                       %Ohms  //Opamp Input resistance for current measurement
inverter.opamp_Gain    = inverter.OpampFb_Rf/inverter.opampInput_R; %Opamp Gain used for current measurement

%% Derive Characteristics
pmsm.N_base = 3125;%mcb_getBaseSpeed(pmsm,inverter); %rpm // Base speed of motor at given Vdc

%% PU System details // Set base values for pu conversion
PU_System = mcb_SetPUSystem(pmsm,inverter);

%% Controller design // Get ballpark values!
% Get PI Gains
PI_params = mcb.internal.SetControllerParameters(pmsm,inverter,PU_System,T_pwm,Ts,Ts_speed);

%Updating delays for simulation
PI_params.delay_Currents  = int32(Ts/Ts_simulink);
PI_params.delay_Position  = int32(Ts/Ts_simulink);
PI_params.delay_Speed     = int32(Ts_speed/Ts_simulink);
PI_params.delay_Speed1    = (PI_params.delay_IIR + 0.5*Ts)/Ts_speed;

%% Open loop reference values
T_Ref_openLoop          = 0.2;                    % Sec // Time for open-loop start-up
Speed_Ref_openLoop      = 300;                  % RPM // Speed referene for open-loop start-up
Iq_Ref_openLoop         = 1;                 % A   // Iq referene for open-loop start-up

%% Gain Calculation
T = Ts*2;
k = (pmsm.Ke/sqrt(3))*2;
num = (k*(1 + (pmsm.Lav + pmsm.Rs*T)/pmsm.Lav));
den = (-pmsm.Rs + 4*(pmsm.Lav + pmsm.Rs*T)/T + pmsm.Rs*(pmsm.Lav + pmsm.Rs*T)/pmsm.Lav);
phi=num/den;
LG= k/phi;